internal/services/manager.go

```diff
@@ -6,35 +6,181 @@ import (
+	return [10]int{8260, 8261, 8262, 8263, 8264, 8265, 8266, 8267, 8268, 8269}
+}
+
+func getAvailablePort(config globalProcessComposeConfig) (int, bool) {
+	ports := defaultPorts()
+	for _, p := range ports {
+		available := true
+		for _, instance := range config.Instances {
+			if instance.Port == p {
+				available = false
+			}
+		}
+		if available {
+			return p, true
+		}
+	}
+	return 0, false
+}
+
+type instance struct {
+	Pid  int `json:"pid"`
+	Port int `json:"port"`
+}
+
+type instanceMap = map[string]instance
+
+type globalProcessComposeConfig struct {
+	Instances instanceMap
+	Path      string   `json:"-"`
+	File      *os.File `json:"-"`
+}
+
+func globalProcessComposeJSONPath() (string, error) {
+	path := xdg.DataSubpath(filepath.Join("devbox/global/"))
+	return filepath.Join(path, "process-compose.json"), errors.WithStack(os.MkdirAll(path, 0755))
+}
+
+func readGlobalProcessComposeJSON(configPath string) globalProcessComposeConfig {
+	config := globalProcessComposeConfig{Instances: map[string]instance{}}
+
+	err := errors.WithStack(cuecfg.ParseFile(configPath, &config.Instances))
+	if err != nil {
+		return config
+	}
+	config.Path = configPath
+	return config
+}
+
+func writeGlobalProcessComposeJSON(config globalProcessComposeConfig) error {
+	// convert config to json using cue
+	json, err := cuecfg.MarshalJSON(config.Instances)
+	if err != nil {
+		return fmt.Errorf("failed to convert config to json: %w", err)
+	}
+
+	// write json to file
+	path, err := globalProcessComposeJSONPath()
+	if err != nil {
+		return fmt.Errorf("failed to get config path: %w", err)
+	}
+
+	if err := os.WriteFile(path, json, 0666); err != nil {
+		return fmt.Errorf("failed to write global config file: %w", err)
+	}
+
+	return nil
+}
+
+func getGlobalConfig() (globalProcessComposeConfig, error) {
+
+	configPath, err := globalProcessComposeJSONPath()
+	if err != nil {
+		return globalProcessComposeConfig{}, fmt.Errorf("failed to get config path: %w", err)
+	}
+
+	globalConfigFile, err := os.OpenFile(configPath, os.O_RDWR|os.O_CREATE, 0666)
+	if err != nil {
+		return globalProcessComposeConfig{}, fmt.Errorf("failed to open config file: %w", err)
+	}
+	lockFile(globalConfigFile.Fd())
+	defer unlockFile(globalConfigFile.Fd())
+
+	config := readGlobalProcessComposeJSON(configPath)
+	config.File = globalConfigFile
+	return config, nil
+}
+
+func addInstance(projectDir string, projectConfig instance) error {
+	config, err := getGlobalConfig()
+	if err != nil {
+		return err
+	}
+
+	lockFile(config.File.Fd())
+	defer unlockFile(config.File.Fd())
+
+	config.Instances[projectDir] = projectConfig
+	err = writeGlobalProcessComposeJSON(config)
+	if err != nil {
+		return fmt.Errorf("failed to write global process-compose config")
+	}
+
+	return nil
+}
+
+func removeInstance(projectDir string) error {
+	config, err := getGlobalConfig()
+	if err != nil {
+		return err
+	}
+
+	lockFile(config.File.Fd())
+	defer config.File.Close()
+	defer unlockFile(config.File.Fd())
```

Comment:
>  -u, --unlock
           Drop a lock. This is usually not required, since a lock is
           automatically dropped when the file is closed.
source: https://man7.org/linux/man-pages/man1/flock.1.html

So, this makes me think that you just want `defer config.File.Close()`.

Reply:
I'd also suggest checking the error return values of `config.File.Close()` and doing:

```
defer func() {
     err := config.File.Close()
     if err  != nil {
         // prefer writer since we manage the --quiet flag using that, but if its hard to get a writer here
         // for now, you could do io.Stderr
          fmt.Fprintf(writer /* or io.Stderr */, "failed to close file (%s). Got error %v\n", someFilePath, err)
     } 

}
```

This way we may get alerted to any errors from conflicts, etc. with our file-locking-and-writing code.

Reply:
also, maybe move the `defer file.config.Close()` to be after the `getGlobalConfig's error check`?

Reply:
Ah that makes sense, if I can unlock the file by closing it, then I can just defer that and only call `unlock` specifically when I need it